Source/WebCore/accessibility/AccessibilityNodeObject.cpp

```diff
@@ -440,8 +440,21 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRoleFromNode(Tr
         return AccessibilityRole::LandmarkMain;
     if (node()->hasTagName(navTag))
         return AccessibilityRole::LandmarkNavigation;
-    if (node()->hasTagName(asideTag))
+    // The aside element should not assume the complementary role when nested
+    // within the sectioning content elements.
+    // https://w3c.github.io/html-aam/#el-aside
+    if (node()->hasTagName(asideTag)) {
+        // Return LandmarkComplementary if the aside element explicitly has a role="complementary".
+        if (equalLettersIgnoringASCIICase(getAttribute(roleAttr), "complementary"_s))
+            return AccessibilityRole::LandmarkComplementary;
+        if (isDescendantOfElementType({ asideTag, articleTag, sectionTag, navTag })) {
+            // Return LandmarkComplementary if the aside element has an accessible name.
+            if (hasAttribute(aria_labelAttr))
```

Comment:
There are a lot of ways something can have an accessible name — `aria-label`, `aria-labelledby`, `aria-description`, `aria-describedby`, more. There are other roles with this requirement, so it would be nice if we had this abstracted in a shared function (e.g. `hasExplicitAccessibleName`).

But I need to think more on that, and don't want to block this PR on it. So for now, maybe we can just check for the presence of any four of those attributes, `hasAttribute(aria_labelAttr) || hasAttribute(aria_labelledby) || ...`? What do you think?

Reply:
Okay,  I will check other aria attributes here like:
```
  if (hasAttribute(aria_labelAttr) || hasAttribute(aria_labelledbyAttr) || hasAttribute(aria_descriptionAttr) || hasAttribute(aria_describedbyAttr))
```
Then, I'll start working on a separate patch to introduce hasExplicitAccessibleName().